### ISSUE TITLE
[analysis_server] Connection to server got closed. Server will restart.

### ISSUE THREAD

**Issue by incendial** [dart-lang/sdk#54441]
Noticed the analysis server silently restarting 9 times within ~5 hours with `true` as the only output message. I'm not sure what that supposed to mean, so let me know if I can provide any additional logs.

<img width="840" alt="Screenshot 2023-12-22 at 11 45 39" src="https://github.com/dart-lang/sdk/assets/21116742/804026cc-16e5-4246-a17d-a337adc3c62a">

What makes these restarts visible is the "your project has some issues that dart fix can fix" message 🙂.

Dart version: Dart SDK version: 3.3.0-174.2.beta

**Comment by bwilkerson** [comment]
Which IDE are you using?

**Comment by incendial** [comment]
VS Code

**Comment by bwilkerson** [comment]
@DanTup Is this an error you recognize?

Based on a quick search, this appears to happen to several of the language servers, not just Dart's. The only "fix" I could find (in https://www.codeproject.com/Tips/5163886/Fixing-VS-Code-Extension-Errors) is to delete the language server's install directory. The implication appears to be that the install directory is sometimes corrupted in a way that VS Code can't recover from.

I'm guessing that the equivalent here would be to delete the DartCode plugin and force it to be downloaded again. I haven't seen the error so I haven't tried the "fix". I don't know how reliable the suggestion is or whether it's safe for you to try it, but that's all the information I have at this point, so I'm passing it along.

**Comment by incendial** [comment]
Well, the server crashes on a hover with pressed cmd (which is a known issue, AFAIK). The problem is that `true` as the output is not very informative.
And I'm not sure that reinstalling the extensions is the solution here.

Btw, for the past 2 days it crashed 15 times...

**Comment by DanTup** [comment]
The error message comes from the VS Code LSP client:

https://github.com/microsoft/vscode-languageserver-node/blob/3f725d8368ec0c75f390a37e4e159d78bc1527ac/client/src/common/client.ts#L1605

This happens when the server terminates (the LSP client is connected to the stdout/stdin streams), so perhaps there's some exception happening here that's causing us to shut the server down.

My first suggestion would be to enable both the [server instrumentation log](https://dartcode.org/docs/logging/#analyzer-instrumentation) and the [client-side log](https://dartcode.org/docs/logging/#analyzer) and see whether any error/exception text is being recorded.

If you have any steps to trigger this I can take a look too (it might depend on your exact SDK version and possibly contents of your file(s) if it's something like an exception during parsing).

(I also suspect reinstalling the extension won't change anything here)

**Comment by DanTup** [comment]
> Well, the server (sometimes) crashes on a hover with pressed cmd (which is a known issue, AFAIK)

I just noticed this... I don't recall this being a known issue (unless it's something fixed and just not shipped yet).


> The problem is that true as the output is not very informative (and it's not clear where it comes from).

Agreed, that's a useless message, though it's not clear to me where it's coming from yet. If you have repro steps I can debug - otherwise the logs noted above might help (it's not clear to me if it's coming from the server or something in the client).

**Comment by jwren** [comment]
I see you are using a beta version of the Dart SDK, `3.3.0-174.2.beta`, have you tried using the stable version?  Having a confirmation that this isn't happening to you on stable, but is happening on beta would be helpful to know.

**Comment by denniskaselow** [comment]
I don't know if it's related, but this `if` https://github.com/dart-lang/sdk/blob/a4e15669eccae7919ed3766c8aafe8bad5a30470/pkg/analyzer/lib/src/dart/analysis/driver.dart#L2368-L2370 is missing a `_hasWork.notify();` or some other way to keep the program running.

Once the while loop arrives at https://github.com/dart-lang/sdk/blob/a4e15669eccae7919ed3766c8aafe8bad5a30470/pkg/analyzer/lib/src/dart/analysis/driver.dart#L2322 after the `continue` the program will silently exit with an exit code 0 because it tries to await the future of a completer that hasn't completed. It can be reproduced with this code:

```dart
import 'dart:async';

void main() async {
  print('creating completer');
  final completer = Completer<void>();
  print('awaiting completer.future');
  await completer.future;
  print('done');
}
```

It'll print 
```
creating completer
awaiting completer.future

Process finished with exit code 0
```

And/Or maybe awaiting `completer.future` shouldn't silently exit the program.

**Comment by DanTup** [comment]
@denniskaselow the reason your example script terminates is that there is nothing keeping the VM alive. That's caught me out too - I thought if `main` returned a `Future` the runtime would stay alive until it completes.

I don't think that applies to the server here though - my expectation is for the server to keep running because there is a listener on `stdin`.  If you add `stdin.listen((_) {});` to your example app, it will no longer terminate (but instead just wait forever at `await completer.future;`).

**Comment by DanTup** [comment]
I think there are two problems here:

1. The analysis server is crashing/terminating
  I think to fix this, we need a log file to understand what the error is
2. The experience when the server crashes in this way is bad
  I've an issue at https://github.com/Dart-Code/Dart-Code/issues/5328 to improve this, replacing the default LSP Client notification with one of our own (that should include the most recent stdout/stderr output for the process)


@incendial do you still see this? Are you able to provide a log?

**Comment by incendial** [comment]
> @incendial do you still see this?

No, I have not seen this crash for a while now (but maybe I'm just not paying attention to it anymore). Feel free to close this.

**Comment by DanTup** [comment]
Thanks - please do ping back (or file a new issue with specific details) if it comes up again!
